[move-only] Implement escaping closure semantics part 1

gottesmm opened the pull request:
This patch series does a bunch of things:

1. Patch 1 changes SILGen/MoveCheckers to emit vars in a boxed eagerly retroject manner.
2. Patch 2 changes "" to also emit lets to be emitted as boxes like vars.
3. Patch 3 changes "" to emit owned arguments again as boxed.
4. Part 4 implements on top of this escaping closure semantics.

This basically required me to rewrite a large part of SILGen for non copyable types and redesign parts of the checker as well. I also had to do a bunch of work in alloc box to stack to massage its output since it semantically is what controls whether we use non-escaping or escaping semantics when we emit these errors.

gottesmm committed:
[gardening] Delete an empty header that snuck in.

gottesmm committed:
[move-only] Update SILGen/MoveCheckers so that vars are emitted in eagerly projected box form.

This is the first slice of bringing up escaping closure support. The support is
based around introducing a new type of SILGen VarLoc: a VarLoc with a box and
without a value. Because the VarLoc only has a box, we have to in SILGen always
eagerly reproject out the address from the box. The reason why I am doing this
is that it makes it easy for the move checker to distinguish in between
different accesses to the box that we want to check separately. As such every
time that we open the box, we insert a mark_must_check
[assignable_but_not_consumable] on that project. If allocbox_to_stack manages to
determine that the box can be stack allocated, we eliminate all of the
mark_must_check and place a new mark_must_check [consumable_and_assignable] on
the alloc_stack.  The end result is that we get the old model that we had before
and also can support escaping closures.

gottesmm committed:
[move-only] Change noncopyable lets to be emitted as boxes like vars.

Some notes:

1. This ensures that if we capture them, we just capture the box by reference.

2. We are still using the old incorrect semantics for captures. I am doing this
   so I can bring this up in separate easy to understand patches all of which
   pass all of the moveonly tests.

3. Most of the test edits are due to small differences in error messages in
   between the object and address checker.

4. I had to add a little support to the move only address checker for a small
   pattern that doesn't occur with vars but do es occur for lets when we codegen
   like this, specifically around enums. The pattern is we perform a load_borrow
   and then copy_value and then use the result of the copy_value. Rather than fight
   SILGen pattern I introduced a small canonicalization into the address checker which
   transforms that pattern into a load [copy] + begin_borrow to restore the codegen
   to a pattern the checker expects.

5. I left noimplicitcopy alone for now. But we should come back around and fix
   it in a similar way. I just did not have time to do so.

gottesmm committed:
[move-only] Box owned arguments like let parameters.

gottesmm committed:
[move-only] Implement escaping closure semantics.

NOTE: A few of the test patterns need to be made better, but this patch series
is large enough, I want to get it into tree and iterate.

gottesmm commented:
@swift-ci smoke test

jckarter commented on test/SILOptimizer/moveonly_nonescaping_closures.swift:
```diff
@@ -46,18 +46,21 @@ func a(x: __shared M) {
     clodger({ borrow(x) })
     borrow(x)
     consume(x)
+    // expected-error @-1 {{'x' was consumed but it is illegal to consume a noncopyable immutable capture of an escaping closure. One can only read from it}}
```
This is not an escaping closure, so this code should be valid without diagnostics.

jckarter commented on lib/SIL/IR/TypeLowering.cpp:
```diff
@@ -112,15 +112,25 @@ CaptureKind TypeConverter::getDeclCaptureKind(CapturedValue capture,
+  // If this is a noncopyable 'let' constant that is not a shared paramdecl,
+  // then we know it is boxed and want to pass it in its boxed form so we can
+  // obey Swift's capture reference semantics.
+  if (!var->supportsMutation() && lowering.getLoweredType().isPureMoveOnly()) {
```
You should also check `!capture.isNoEscape()` here so we don't box captures from nonescaping closures.

gottesmm replied:
I am trying that now

gottesmm committed:
[move-only] Fix IRGen deinit test for stack changes.

gottesmm committed:
[move-only] Do not box captures if the capturing closure is no escape.

Thanks to JoeG for seeing this quickly and preventing me from having to track
this down!

gottesmm committed:
[move-only] Suppress moveonly errors when an inout is captured by an escaping closure since we will already emit a more specific capture error.

gottesmm commented:
@swift-ci smoke test

gottesmm commented:
@swift-ci build toolchain

gottesmm commented:
Linux failure was unrelated problem in dispatch.

gottesmm commented:
@swift-ci smoke test linux platform

gottesmm committed:
Remove some optnone that snuck in.

gottesmm commented:
Windows failed b/c I left in some optnone by mistake from when I was debugging

gottesmm commented:
@swift-ci smoke test

gottesmm commented:
@swift-ci build toolchain

gottesmm commented:
@swift-ci clean smoke test linux platform
